lambda/translcore.ml

```diff
@@ -1113,15 +1114,18 @@ and transl_function0
             (value_kind pat.pat_env pat.pat_type))
           (value_kind pat.pat_env pat.pat_type) other_cases
     in
+    let body =
+      Matching.for_function ~scopes loc repr (Lvar param)
+        (transl_cases ~scopes cases) partial
+    in
+    let region = region || not (may_allocate_in_region body) in
```

Comment:
This is a bit tricky. There are two separate bits of information about a function:

  - `lfunction.region`: if true, this function is known not to allocate into its caller's region
  - `Lregion` around body: prevents local allocations in the body from escaping into its caller's region

An ordinary function does not allocate into its caller's region, so has `lfunction.region = true` and wraps the body in an `Lregion`. A local-returning function, on the other hand, has `lfunction.region = false` and does not wrap the body in an `Lregion`.

We previously implemented an optimisation to the above:

  - If an ordinary function contains no locally-allocating operations, then the `Lregion` may be elided even though `lfunction.region = true` - we need no `Lregion` to prevent escape if there is nothing that could escape in the first place.

This PR adds the dual optimisation:

  - If a local-returning function contains no locally-allocating operations, then `lfunction.region` may be marked `true` even though there is no `Lregion` - again, nothing can escape if there is nothing that could escape in the first place.

This dual optimisation is important for a specific edge case of functions: the partial application stubs created by the curry-disabling effect of mutable patterns in function parameters. Without this optimisation, these functions get `lfunction.region = false`, even when they do not contain any local allocations. This means that overapplications of these functions (the normal case) see an overapplication of an `lfunction.region = false` function, requiring them to introduce a region and breaking tail call optimisation.